internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,371 @@
+package graph
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/oklog/ulid/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"go.uber.org/goleak"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/storage/memory"
+	"github.com/openfga/openfga/pkg/testutils"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+func TestSingleflightCheckResolver_TwoProngLoop(t *testing.T) {
+	storeID := ulid.Make().String()
+
+	ds := memory.New()
+	err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:1", "owner", "user:anne"),
+		tuple.NewTupleKey("document:2", "parent", "document:1"),
+		tuple.NewTupleKey("document:1", "parent", "document:2"),
+	})
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+  schema 1.1
+
+type user
+type document
+  relations
+	define parent: [document]
+	define owner: [user] or owner from parent
+    define viewer: owner or viewer from parent`)
+
+	err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
+	require.NoError(t, err)
+
+	ctx := typesystem.ContextWithTypesystem(
+		context.Background(),
+		typesystem.New(model),
+	)
+	ctx = storage.ContextWithRelationshipTupleReader(ctx, ds)
+
+	singleflightCheckResolver := NewSingleflightCheckResolver()
+	localCheckResolver := NewLocalChecker()
```

Comment:
These need the `CycleDetectionCheckResolver`. My recommendation for the test here is to use the `NewLayeredCheckResolver(nil, false, nil)` so you get the whole layered resolution chain (without the caching).